File: app/models/quiz.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Quiz represents a quiz in the application.
type Quiz struct {
	ID                    primitive.ObjectID `json:"id,omitempty" bson:"_id,omitempty"`
	Title                 string             `json:"title" bson:"title"`
	Sections              []string           `json:"sections" bson:"sections"`
	Difficulty            string             `json:"difficulty" bson:"difficulty"`
	Questions             []Question         `json:"questions" bson:"questions"`
	CollectEmail          bool               `json:"collect_email" bson:"collect_email"`
	AllowMultipleAttempts bool               `json:"allow_multiple_attempts" bson:"allow_multiple_attempts"`
	Active                bool               `json:"active" bson:"active"`
	Timer                 int                `json:"timer" bson:"timer"`
	RandomiseQuestions    bool               `json:"randomise_questions" bson:"randomise_questions"`
	CreatedAt             time.Time          `json:"created_at" bson:"created_at"`
	UpdatedAt             time.Time          `json:"updated_at" bson:"updated_at"`
	CreatedBy             string             `json:"created_by" bson:"created_by"`
}

type QuizStats struct {
	NumberOfAttempts       int     `json:"attempts"`
	AverageScore           float64 `json:"average_score"`
	AverageAttemptsPerUser float64 `json:"average_attempts_per_user"`
}

File: web/app/types/quiz.ts
export type Quiz = {
    id?: string;
    title: string;
    sections: string[];
    difficulty: string;
    questions: Question[];
    collect_email: boolean;
    allow_multiple_attempts: boolean;
    created_at?: string;
    updated_at?: string;
    created_by: string;
};

export type Question = {
    id?: string;
    text: string;
    choices: string[];
    correct: string;
    difficulty: string;
    section: string;
};

export type QuizAnswer = {
    id?: string;
    quiz_id: string;
    user_id: string;
    answers: QuestionAnswer[];
    started_at?: string;
    ended_at?: string;
    created_at?: string;
}

export type QuestionAnswer = {
    question_id: string;
    answer: string;
}

export type Stats = {
    attempts: number;
    average_score: number;
    average_attempts_per_user: number;
}